### PR TITLE
AWS Redshift: fix parameters check and port module to boto3

### LINKED ISSUE #37051: Redshift module issues on 2.4 and devel
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

redshift
##### ANSIBLE VERSION

```
$ ansible --version
ansible 2.6.0 (devel e3895e4d83) last updated 2018/03/05 21:53:15 (GMT -300)
  config file = None
  configured module search path = [u'/home/rdriutti/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/rdriutti/code/ansible/lib/ansible
  executable location = /home/rdriutti/code/ansible/bin/ansible
  python version = 2.7.14 (default, Sep 23 2017, 22:06:14) [GCC 7.2.0]

```
```
22:10 $ ansible --version
ansible 2.4.3.0
  config file = None
  configured module search path = [u'/home/rdriutti/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/rdriutti/.local/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.14 (default, Sep 23 2017, 22:06:14) [GCC 7.2.0]
```
##### OS / ENVIRONMENT

Ubuntu 17.10
##### SUMMARY

The redshift module fails to create a cluster because of missing parameters that are required only when the command is *delete*. The logic to check the parameters is wrong, shouldn't care for  `skip_final_cluster_snapshot` or `final_cluster_snapshot_identifier` when the command is create or modifiy.
If using the devel branch, It also fails when creating a cluster because it tries to use a feature available only on boto3, but the module claims it's able to run with boto 2.9+, but as of boto 2.48, using `enhanced_vpc_routing` on a CreateCluster call will throw an exception

##### STEPS TO REPRODUCE

Run the playbook on localhost using the keys for some aws account you have access:

```yaml
---

- hosts: localhost
  tasks:

  - name: Launch Redshift 
    local_action:
      module: redshift
      command: create 
      cluster_type: "single-node"
      db_name: "rafa-test"
      identifier: "rafa-redshift"
      maint_window: "Mon:02:00-Mon:02:30"
      node_type: "dc1.large"
      param_group_name: "default.redshift-1.0"
      region: "us-east-1"
      username: master
      password: "asdfqwerAA11"
      automated_snapshot_retention_period: 0
      wait: true
      wait_timeout: 3000

```

##### EXPECTED RESULTS

Redshift cluster creted

##### ACTUAL RESULTS

Got an error

This is the params validation error (2.4.8 and devel)
```
TASK [Launch Redshift instances] ******************************************************************************************************************************************************************************
fatal: [localhost -> localhost]: FAILED! => {"changed": false, "msg": "skip_final_cluster_snapshot is False but all of the following are missing: final_cluster_snapshot_identifier"}
        to retry, use: --limit @/home/rdriutti/code/ansible/create-cluster.retry

PLAY RECAP ****************************************************************************************************************************************************************************************************
localhost  
```
This is only on devel:
```
TASK [Launch Redshift instances] ******************************************************************************************************************************************************************************
task path: /home/rdriutti/code/ansible/create-cluster.yml:6
Using module file /home/rdriutti/code/ansible/lib/ansible/modules/cloud/amazon/redshift.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: rdriutti
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/rdriutti/.ansible/tmp/ansible-tmp-1520300144.74-278909426876762 `" && echo ansible-tmp-1520300144.74-278909426876762="` echo /home/rdriutti/.ansible/tmp/ansible-tmp-1520300144.74-278909426876762 `" ) && sleep 0'
<localhost> PUT /home/rdriutti/.ansible/tmp/ansible-local-18151rOE_gi/tmpX7AGbv TO /home/rdriutti/.ansible/tmp/ansible-tmp-1520300144.74-278909426876762/redshift.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/rdriutti/.ansible/tmp/ansible-tmp-1520300144.74-278909426876762/ /home/rdriutti/.ansible/tmp/ansible-tmp-1520300144.74-278909426876762/redshift.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/rdriutti/.ansible/tmp/ansible-tmp-1520300144.74-278909426876762/redshift.py && sleep 0'
<localhost> EXEC /bin/sh -c 'rm -f -r /home/rdriutti/.ansible/tmp/ansible-tmp-1520300144.74-278909426876762/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_QpwK2x/ansible_module_redshift.py", line 515, in <module>
    main()
  File "/tmp/ansible_QpwK2x/ansible_module_redshift.py", line 500, in main
    (changed, cluster) = create_cluster(module, conn)
  File "/tmp/ansible_QpwK2x/ansible_module_redshift.py", line 298, in create_cluster
    redshift.create_cluster(identifier, node_type, username, password, **params)
TypeError: create_cluster() got an unexpected keyword argument 'enhanced_vpc_routing'

fatal: [localhost -> localhost]: FAILED! => {
    "changed": false, 
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_QpwK2x/ansible_module_redshift.py\", line 515, in <module>\n    main()\n  File \"/tmp/ansible_QpwK2x/ansible_module_redshift.py\", line 500, in main\n    (changed, cluster) = create_cluster(module, conn)\n  File \"/tmp/ansible_QpwK2x/ansible_module_redshift.py\", line 298, in create_cluster\n    redshift.create_cluster(identifier, node_type, username, password, **params)\nTypeError: create_cluster() got an unexpected keyword argument 'enhanced_vpc_routing'\n", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE", 
    "rc": 1
}

```

### PR DESCRIPTION
**Pull request by rafaeldriutti**
##### SUMMARY

Make it work with current ansible and boto

Fixes #37051

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

redshift
##### ANSIBLE VERSION

```
$ ansible --version
ansible 2.6.0 (devel e3895e4d83) last updated 2018/03/05 21:53:15 (GMT -300)
  config file = None
  configured module search path = [u'/home/rdriutti/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/rdriutti/code/ansible/lib/ansible
  executable location = /home/rdriutti/code/ansible/bin/ansible
  python version = 2.7.14 (default, Sep 23 2017, 22:06:14) [GCC 7.2.0]
```

##### ADDITIONAL INFORMATION

See https://github.com/boto/boto/blob/2.48.0/boto/redshift/layer1.py#L313 for the enhanced_vpc_routing thing.
More info on the bug report.

### CODE REVIEW

**Comment by willthames** [comment]
This change will require a full test suite as per our policy: https://github.com/ansible/community/blob/master/group-aws/integration.md#policy

Documentation on providing a test suite is at: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md#integration-tests-for-aws-modules

Let us know if there's any help we can provide. A good starting model is the `ec2_group` test suite (test/integration/targets/ec2_group)

**Comment by mattclay** [comment]
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/58254/66/tests

File: lib/ansible/modules/cloud/amazon/redshift.py
Comment (psharkey) [inline lib/ansible/modules/cloud/amazon/redshift.py]: Should this be: `params['d_b_name'] = d_b_name`?

File: test/integration/targets/redshift/tasks/main.yml
Comment (s-hertel) [inline test/integration/targets/redshift/tasks/main.yml]: Is there a reason this one is commented out?

File: test/integration/targets/redshift/tasks/main.yml
Comment (s-hertel) [inline test/integration/targets/redshift/tasks/main.yml]: s/clenup/cleanup

File: lib/ansible/modules/cloud/amazon/redshift.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/redshift.py]: I know this isn't something you changed, but this should wait until it is finished deleting. Do you mind fixing it? This will allow the tests to pass with tasks after deleting a cluster without relying on pause and retries.

File: lib/ansible/modules/cloud/amazon/redshift.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/redshift.py]: The connection does not need any exception handling. Also, if you rebase on devel, instead of calling get_aws_connection_info and boto3_conn, remove those imports and just use `conn = module.client('redshift')`

**Review by s-hertel — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/redshift.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/redshift.py]: This should catch only the specific error that occurs when the cluster doesn't exist yet. You can import is_boto3_error_code from module_utils.aws.core and do: `except is_boto3_error_code('ClusterNotFound'):`

File: lib/ansible/modules/cloud/amazon/redshift.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/redshift.py:350]: There should be an `except is_boto3_error_code('ClusterNotFound'):` here that would return (False, {}) for idempotence if the cluster doesn't exist.

**Review by s-hertel — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/redshift.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/redshift.py:476]: You can remove some connection boilerplate (`get_aws_connection_info`, `if not region:`,  `boto3_conn(..`) and replace it with `conn = module.client('redshift')`

File: lib/ansible/modules/cloud/amazon/redshift.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/redshift.py]: type='str' is the default so you can omit it if you want. Typically the non-defaults are listed for AWS modules.

File: lib/ansible/modules/cloud/amazon/redshift.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/redshift.py]: Rather than doing `if p in module.params:` on [line 283](https://github.com/ansible/ansible/pull/37052/files#diff-fd4728adea42973936ca9bdb0082deadR283) you could move this line `if module.params.get(p) is not None:` there and unindent the line below.

File: lib/ansible/modules/cloud/amazon/redshift.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/redshift.py]: Should retrieving the resource happen after waiting if wait is True? It seems like this could fail sometimes.

File: lib/ansible/modules/cloud/amazon/redshift.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/redshift.py:421]: I'm not sure why an error describing the cluster leads to trying to modify the cluster. What is the exception expected [here](https://github.com/ansible/ansible/pull/37052/files#diff-fd4728adea42973936ca9bdb0082deadR409)?

File: lib/ansible/modules/cloud/amazon/redshift.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/redshift.py]: "Failed to describe cluster" would be a more accurate error message.

File: lib/ansible/modules/cloud/amazon/redshift.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/redshift.py:249]: Should you default these values to None that way the returned keys are always the same? It would also be good to add this note to the return docs for these four.

File: test/integration/targets/redshift/aliases
Comment (s-hertel) [inline test/integration/targets/redshift/aliases:3]: I think these take too long to run in CI... 25 minutes and counting. They should probably have the `unsupported` alias added and people will need to run them manually before merging redshift things.

Update - they took 34 minutes, which is not the worst but is still probably too slow.

**Review by s-hertel — COMMENTED** [review]
@rafaeldriutti Thanks for all the work you've put into this, it's looking good! Feel free to add yourself to the author list if you'd like to help maintain and merge things from the community for the module in the future.

File: test/integration/targets/redshift/aliases
Comment (rafaeldriutti) [reply test/integration/targets/redshift/aliases:3]: I can make it faster by removing the multi-node creation and delete test, that adds like 10-15 mins to the run.

**Comment by rafaeldriutti** [comment]
How can I add the ModifyCluster permission to the CI policy?
`An error occurred (AccessDenied) when calling the ModifyCluster operation: User: arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=shippable=ansible=ansible=84814.66 is not authorized to perform: redshift:ModifyCluster on resource: arn:aws:redshift:us-east-1:966509639900:cluster:shippable-84814-66"`

**Comment by mattclay** [comment]
@rafaeldriutti Adding permissions to CI is something that someone on the Ansible Core Team will need to do. This PR is labeled `needs_ci_update`, so it's on the list of PRs which are blocked on CI permissions.

**Review by s-hertel — APPROVED** [review]

**Comment by rafaeldriutti** [comment]
Thanks for all the help @s-hertel :+1:  Seems the linter doesn't like the additional exeption handling added on b96675f.

**Comment by rafaeldriutti** [comment]
Should I just remove the modify cluster test?

**Comment by s-hertel** [comment]
bot_status

**Comment by s-hertel** [comment]
Sorry for the delay, @rafaeldriutti. We hit the size limit on the policy for CI and some things had to be rearranged before fixing the permissions was possible. Thank you for the time you spent on this!

**Pull request merged**
